Add Quick start section on dispatching events (#1692)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/Hooks/EventDispatcher/Index.rst

.. _EventDispatcherQuickStart:

Quick start
===========

.. _EventDispatcherQuickStartDispatching:

Dispatching an event
--------------------

#. Create an event class. 

   An event class is basically a plain PHP object with getters for immutable
   properties and setters for mutable properties. It contains a constructor for all properties::
   
      final class DoingThisAndThatEvent {
         private string mutableProperty;
         private int immutableProperty;
         
         public function __construct(string mutableProperty, int immutableProperty) {
            // ...
         }
         
         // Getter for both properties, setters only for $mutableProperty;
      }
      
   Read more about :ref:`implementing event classes <EventDispatcherEvents>`.
      
#. Inject the EventDispatcher

   If you are in a controller the `EventDispatcher` already got injected
   and in this case you can omit this step.
   
   If the EventDispatcher is not yet available, you need have it injected::
   
      use Psr\EventDispatcher\EventDispatcherInterface;
      
      final class MyClass {

>             

Reviewer: ```suggestion
```

          private EventDispatcherInterface $eventDispatcher;
          
          public function injectEventDispatcher(EventDispatcherInterface $eventDispatcher): void
          {
              $this->eventDispatcher = $eventDispatcher;
          }
      }
      
#. Dispatch the event

   Create an event object with the data that should be passed to the listeners. Use the data 
   of mutable properties however it suits your business logic::
   

>    public doSomething() {

Reviewer: ```suggestion
   public function doSomething() {
```

Can you also fix the indentation of the code block? ;-)

>          // ..
>          /** @var DoingThisAndThatEvent $event */
>          $event = $this->eventDispatcher->dispatch(
>             new DoingThisAndThatEvent("foo", "bar")
>          );
>          $someChangedValue = $event->getMutableProperty();
>          // ...

Reviewer: ```suggestion
          // ..
          /** @var DoingThisAndThatEvent $event */
          $event = $this->eventDispatcher->dispatch(
              new DoingThisAndThatEvent("foo", "bar")
          );
          $someChangedValue = $event->getMutableProperty();
          // ...
```

  Additional background on the implementation can be found at https://usetypo3.com/psr-14-events.html

.. _EventDispatcherQuickStart:

Quick start
===========

.. _EventDispatcherQuickStartDispatching:

Dispatching an event
--------------------

#. Create an event class. 

   An event class is basically a plain PHP object with getters for immutable
   properties and setters for mutable properties. It contains a constructor for all properties::
   

>       class DoingThisAndThatEvent {

Reviewer: ```suggestion
      final class DoingThisAndThatEvent {
```

All new code should be [final by default](https://matthiasnoback.nl/2018/09/final-classes-by-default-why/).

         private string mutableProperty;

>          private int inmutableProperty;

Reviewer: ```suggestion
         private int immutableProperty;
```

>          public function __construct(string mutableProperty, int inmutableProperty) {

Reviewer: ```suggestion
         public function __construct(string mutableProperty, int immutableProperty) {
```

            // ...
         }
         
         // Getter for both properties, setters only for $mutableProperty;
      }
      
   Read more about :ref:`implementing event classes <EventDispatcherEvents>`.
      
#. Inject the EventDispatcher

>    If you are in a Controller the EventDispatcher already got injected in the ActionController

Reviewer: ```suggestion
   If you are in a controller the `EventDispatcher` already got injected
```

>    in this case you can ommit this step.

Reviewer: ```suggestion
   and in this case you can omit this step.
```

>    If the EventDispatcher is not yet available, you need inject it::

Reviewer: ```suggestion
   If the EventDispatcher is not yet available, you need have it injected::
```

>       class MyClass {

Reviewer: ```suggestion
      final class MyClass {
```

>           use Psr\EventDispatcher\EventDispatcherInterface;

Reviewer: This must be moved above.

>           /**
>            * @var EventDispatcherInterface
>            */
>           private $eventDispatcher;

Reviewer: ```suggestion
          private EventDispatcherInterface $eventDispatcher;
```

          
          public function injectEventDispatcher(EventDispatcherInterface $eventDispatcher): void
          {
              $this->eventDispatcher = $eventDispatcher;
          }
      }
      
#. Dispatch the event

   Create an event object with the data that should be passed to the listeners. Use the data 
   of mutable properties however it suits your business logic::
   
   public doSomething() {
      // ..
      /** @var DoingThisAndThatEvent $event */
      $event = $this->eventDispatcher->dispatch(
         new DoingThisAndThatEvent("foo", "bar")
      );

>       $someChangedValue = $event->getMutableProperty();

Reviewer: We should explicitly add this getter (and a setter) to the event above.

Author: I think every developer should know what a getter and setter look like. Most ides generate them automatically. And it would only blow up the quick start example

Reviewer: Indeed. I usually just like to have everything explicit so that even the copypasta people have some success. ;-)
